Use single threaded executor in benchmarks and tests

A-Walrus opened the pull request:
# Objective
Fixes: #9838
## Solution
Add `single_threaded` constructor for `Schedule` and use it in all benchmarks and tests.

A-Walrus committed:
Switch benchmarks to single_threaded executor

A-Walrus committed:
Move tests to `single_threaded`

A-Walrus committed:
Fix  tests requiring `MultiThreaded` executor

hymm commented:
You shouldn't make this change for any of those scheduling benchmarks. Those are specifically measuring multithreaded execution.

hymm commented on crates/bevy_ecs/src/schedule/mod.rs:
```diff
@@ -125,7 +125,7 @@ mod tests {
         #[test]
         fn order_systems() {
```
I think all the tests that are using checking the SystemOrder with multiple systems should remain multithreaded. They're more likely to fail in the multithreaded scenario.

i.e. test that have assert_eq!(world.resource::<SystemOrder>().0, vec![with multiple elements here]);

hymm reviewed (commented):
Can you also do a pass over the doc examples? Those get run during testing too. A lot of them are using Schedule::default.

A-Walrus commented on crates/bevy_ecs/src/query/mod.rs:
```diff
@@ -791,7 +791,7 @@ mod tests {
         }
 
-        let mut schedule = Schedule::default();
+        let mut schedule = Schedule::default(); // Uses MultiThreaded executor
```
When I tried it with the singlethreaded schedule it didn't work... might have accidentaly uncovered a bug :P

A-Walrus replied:
Nevermind, changed it to `single_threaded` and it seems to work fine :P

A-Walrus replied:
```
---- query::tests::par_iter_mut_change_detection stdout ----
thread 'query::tests::par_iter_mut_change_detection' panicked at 'A ComputeTaskPool has not been initialized yet. Please call ComputeTaskPool::init beforehand.', crates/bevy_tasks/src/usages.rs:28:33
Encountered a panic in system `bevy_ecs::query::tests::par_iter_mut_change_detection::propagate_system`!


failures:
    query::tests::par_iter_mut_change_detection
```
Got a failure for this test on the latest commit. Might be flaky...

A-Walrus committed:
Revert change for Schedule related benches/tests

A-Walrus committed:
Change to single_threaded in doc examples

Left the visible ones as default in order to not make things confusing.
The invisible ones (marked with `/// #`) I changed to single_threaded.

hymm commented on benches/benches/bevy_ecs/components/archetype_updates.rs:
```diff
@@ -12,8 +12,7 @@ struct A<const N: u16>(f32);
 fn setup(system_count: usize) -> (World, Schedule) {
```
there's a warning about an unused variable: `warning: unused import: `ExecutorKind``

hymm commented:
I think we should make a `multithreaded` method that just wraps `default` to make it explicit we're choosing the multithreaded executor instead of leaving a comment.

JMS55 commented:
Same for rendering benchmarks, we want those to be multithreaded.

A-Walrus committed:
Add `multi_threaded` variant and fixup usage

A-Walrus commented:
> Same for rendering benchmarks, we want those to be multithreaded.

Which rendering benchmarks are you referring to? I couldn't find any in the `benches` dir/

JMS55 commented:
Ah nvm then. All the rendering "benchmarks" are actually examples, not part of benches.

hymm commented on crates/bevy_ecs/src/schedule/schedule.rs:
```diff
@@ -190,6 +194,30 @@ impl Schedule {
+    /// [`MultiThreaded`]: ExecutorKind::MultiThreaded
+    pub fn multi_threaded() -> Self {
+        let mut schedule = Self::default();
+        schedule.set_executor_kind(ExecutorKind::SingleThreaded);
```
this should be MultiThreaded

A-Walrus committed:
Fix copy-paste error in multi_threaded constructor

alice-i-cecile reviewed (approved):
This looks like the right setup now to me.

alice-i-cecile commented:
@A-Walrus that CI failure looks like it might be real: can you reproduce it locally?

A-Walrus commented:
@alice-i-cecile It's passing locally, but it has failed in CI multiple times. Weird

hymm commented:
So we should either do like the error is saying and call `ComputeTaskPool::init()` before running the test or change `par_for_each_unchecked_manual` to call `ComputeTaskPool::init()`. I suspect the former is more correct, since that gives the user a chance to init the task pool how they like. We do call `init` when we're using the multithreaded executor. @james7132 do you have any opinion on this?
